src/object/mod.rs

```
@@ -0,0 +1,168 @@
+    password: SecretString,
+}
+
+impl Credentials {
+    /// Create new credentials object
+    pub fn create<'e, 'p>(
+        email: impl Into<Cow<'e, str>>,
+        password: impl Into<Cow<'p, str>>,
+    ) -> Result<Credentials, email_address::Error> {
+        Ok(Credentials {
+            email: EmailAddress::from_str(&email.into())?,
+            password: Secret::new(password.into().to_string()),
+        })
+    }
+}
+
+mod serde_utils {
+    use html_parser::Dom;
+    use serde::{Deserialize, Deserializer};
+    use time::{Duration, OffsetDateTime};
+
+    pub fn duration_from_seconds<'de, D: Deserializer<'de>>(
+        deserializer: D,
+    ) -> Result<Duration, D::Error> {
+        let integer: i64 = Deserialize::deserialize(deserializer)?;
+        Ok(Duration::seconds(integer))
+    }
+
+    // I have no idea how and why "isLiked" field is represented by a number on a site.
+    // This is weird
+    pub fn flag_from_number<'de, D: Deserializer<'de>>(deserializer: D) -> Result<bool, D::Error> {
+        let num: usize = Deserialize::deserialize(deserializer)?;
+        Ok(num > 0)
+    }
+
+    pub fn html_from_str<'de, D: ::serde::de::Deserializer<'de>>(
+        deserializer: D,
+    ) -> Result<Dom, D::Error> {
+        use ::serde::de::Error;
+        let s: String = serde::Deserialize::deserialize(deserializer)?; // FIXME &str results in an error, for some reason
+        let s = s.as_str();
+        Dom::parse(s).map_err(|_| {
+            D::Error::invalid_value(::serde::de::Unexpected::Str(s), &"Valid html fragment")
+        })
+    }
+
+    #[allow(clippy::unnecessary_wraps)]
+    pub fn optional_iso_time<'de, D: ::serde::de::Deserializer<'de>>(
+        deserializer: D,
+    ) -> Result<Option<OffsetDateTime>, D::Error> {
+        match Option::<OffsetDateTime>::deserialize(deserializer) {
+            Ok(option) => Ok(option),
+            Err(_) => Ok(None), // FIXME I guess, that's not the best practice, definitely should redo that
```

Maybe just return the deserialization error as is?

---

As far as I understand it, it would produce an error for this field missing. Which is not what I want (this deserialization function is used for a struct that **might** have a date missing :idk:)